### PR TITLE
Add support for server-side request cancellation

### PR DESCRIPTION
**Pull request by bryce-anderson**
Problem

Performing some server computations may be pretty expensive and in the case where the channel gets shutdown, either because the client hung up or the H2 stream was cancelled (due to a timeout, for example), the server will continue that processing in vein.

Solution

Propagate cancellation into the service handling logic in case it can save itself some work or potentially notify its own backends.

### CODE REVIEW

File: server/src/test/scala/org/http4s/netty/server/ServerTest.scala [hunk 1]
```diff
@@ -117,6 +120,38 @@ abstract class ServerTest extends IOSuite {
       }
     }
   }
+
+  test("requests can be cancelled") {
+    val cancelled = Promise[Unit]()
+    val doCancel = IO.delay {
+      cancelled.success(())
+      ()
+    }
```
Comment (bryce-anderson) [inline server/src/test/scala/org/http4s/netty/server/ServerTest.scala:129]: This feels non-idiomatic. I'm happy to take suggestions to improve it.

Comment (hamnis) [reply server/src/test/scala/org/http4s/netty/server/ServerTest.scala:129]: You can do something similar we do in DrainResponseTest with Defferred and IO.race

Comment (bryce-anderson) [reply server/src/test/scala/org/http4s/netty/server/ServerTest.scala:129]: Thanks, I'll take a look.

Comment (bryce-anderson) [reply server/src/test/scala/org/http4s/netty/server/ServerTest.scala]: That's a bit cleaner, thank you.

**Pull request merged**